### PR TITLE
fix: DHIS2-10001 logic added to tracker importer to validate TEIs mandatory attributes

### PR DESCRIPTION
**Pull request by gnespolino**

### CODE REVIEW

**Review by luciano-fiandesio — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java [hunk 1]
```diff
@@ -97,6 +104,47 @@ public void validateTrackedEntity( ValidationErrorReporter reporter, TrackedEnti
...
+    }
+
+    private void validateMandatoryAttributes( ValidationErrorReporter reporter, TrackedEntity trackedEntity )
+    {
+        TrackedEntityType trackedEntityType = reporter.getValidationContext()
+            .getTrackedEntityType( trackedEntity.getTrackedEntityType() );
+
+        if ( trackedEntityType != null )
+        {
+            Set<String> trackedEntityAttributes = Optional.of( trackedEntity )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java:117]: This TEI passed to validation cannot be null, so we could simply get the attributes without the optional

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java [hunk 2]
```diff
@@ -97,6 +104,47 @@ public void validateTrackedEntity( ValidationErrorReporter reporter, TrackedEnti
...
+                .map( TrackedEntity::getAttributes )
+                .orElse( Collections.emptyList() )
+                .stream()
+                .map( Attribute::getAttribute )
+                .collect( Collectors.toSet() );
+
+            trackedEntityType.getTrackedEntityTypeAttributes()
+                .stream()
+                .filter( this::isMandatory )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java:126]: I found difficult following this Optional.of and then orElse flow and I am not sure if in this case you could really have a null TrackedEntityTypeAttribute in the list of trackedEntityType.getTrackedEntityTypeAttributes()
I would go with 

```suggestion
               .filter(Obects::isNotNull)
               .filter(TrackedEntityTypeAttribute::isMandatory)
```

but of course this is purely personal style

**Review by enricocolasante — APPROVED** [review]
I just put 2 style suggestions

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java (hunk 1 above)
Comment (gnespolino) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java:117]: you're right, thank you

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java (hunk 2 above)
Comment (gnespolino) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/TrackedEntityAttributeValidationHook.java:126]: yes, your version is much easier, I'll change it

**Pull request merged**